### ISSUE TITLE
False positive cyclic-import

### ISSUE THREAD

**Issue by andy-maier** [PyCQA/pylint#5168]
### Bug description

Sorry for just linking our pylint run on the actual project. I tried to come up with a simplified set of files that reproduce the problem, but gave up.

The problem is that pylint 2.11 reports multiple "cyclic-import" issues on our zhmccli project.

Github actions run: https://github.com/zhmcclient/zhmccli/runs/3914601838?check_suite_focus=true

Pylint output:

```
$ pylint --disable=fixme --rcfile=.pylintrc --output-format=text setup.py zhmccli/_cmd_adapter.py zhmccli/_cmd_storagevolume.py zhmccli/zhmccli.py zhmccli/_cmd_info.py zhmccli/_cmd_partition.py zhmccli/_cmd_capacitygroup.py zhmccli/_cmd_hba.py zhmccli/_version.py zhmccli/_cmd_cpc.py zhmccli/_cmd_vfunction.py zhmccli/_cmd_lpar.py zhmccli/__init__.py zhmccli/_cmd_vswitch.py zhmccli/_cmd_vstorageresource.py zhmccli/_cmd_port.py zhmccli/_helper.py zhmccli/_cmd_metrics.py zhmccli/_cmd_nic.py zhmccli/_cmd_storagegroup.py zhmccli/_cmd_session.py   tests/__init__.py tests/function/test_info.py tests/function/utils.py tests/function/__init__.py tests/function/test_global_opts.py   
************* Module tests.function.test_global_opts
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_adapter) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vfunction) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_storagevolume) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_info) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vfunction -> zhmccli._cmd_partition) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_capacitygroup) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_hba) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_adapter -> zhmccli._cmd_cpc) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_lpar) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vswitch) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vstorageresource) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vfunction -> zhmccli._cmd_partition -> zhmccli._cmd_storagegroup -> zhmccli._cmd_port -> zhmccli._cmd_adapter) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vfunction -> zhmccli._cmd_partition -> zhmccli._cmd_metrics) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_nic) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_vfunction -> zhmccli._cmd_partition -> zhmccli._cmd_storagegroup) (cyclic-import)
tests/function/test_global_opts.py:1:0: R0401: Cyclic import (zhmccli -> zhmccli._cmd_session) (cyclic-import)

-----------------------------------
Your code has been rated at 9.95/10
```

This happens only starting with pylint 2.11.0. With pylint 2.10.0 this is not reported.

When running the zhmccli.cli script and when running the tests, Python loads the modules just fine.

I suspect that this is a false positive introduced in pylint 2.11.0

### Command used

```shell
See above
```

### Pylint output

```shell
See above
```

### Expected behavior

Pylint should not report cyclic-import.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.2
Python 3.9.7 (default, Sep  3 2021, 12:45:31) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

### OS / Environment

macOS 11.6

### Additional dependencies

$ pipdeptree -p pylint
pylint==2.11.1
  - astroid [required: >=2.8.0,<2.9, installed: 2.8.2]
    - lazy-object-proxy [required: >=1.4.0, installed: 1.6.0]
    - setuptools [required: >=20.0, installed: 58.2.0]
    - typing-extensions [required: >=3.10, installed: 3.10.0.2]
    - wrapt [required: >=1.11,<1.13, installed: 1.12.1]
  - isort [required: >=4.2.5,<6, installed: 5.9.3]
  - mccabe [required: >=0.6,<0.7, installed: 0.6.1]
  - platformdirs [required: >=2.2.0, installed: 2.4.0]
  - toml [required: >=0.7.1, installed: 0.10.2]
  - typing-extensions [required: >=3.10.0, installed: 3.10.0.2]

**Comment by Pierre-Sassoulas** [comment]
@Cielquan There's a real cyclic import in the file so it's not a false positive. I think a stackoverflow question is better suited to ask for how to redesign the code so you don't have a cyclic import anymore.

**Comment by Cielquan** [comment]
I am reexporting all the models in the sub-package's `__init__.py` anyways so I could call `update_forward_refs` there. The `update_forward_refs` calls are only there to resolve the stringified type annotations.

The problem I see is that it would still need the cyclic imports. Reason is that `User` still needs to be imported to `user_role.py` to be available for the type annotation and vice versa. And those imports must be available at runtime so a "hidden" typing import does unfortunately not work.

I appreciate your help thou.

**Comment by Cielquan** [comment]
I can move the `update_forward_refs` to `__init__.py` -- no problem. But as soon as I hide the cyclic imports behind `if TYPE_CHECKING` I get a NameError like `NameError: name 'User' is not defined` because the model in the type annotations is not defined.
